crates/phactory/src/system/side_tasks/geo_probe.rs

```diff
@@ -0,0 +1,158 @@
+use chain::BlockNumber;
+use phala_mq::Sr25519MessageChannel;
+
+use crate::side_task::async_side_task::AsyncSideTask;
+use crate::side_task::SideTaskManager;
+
+use std::str::FromStr;
+use phala_types::contract;
+use phala_types::messaging::{Geocoding, GeolocationCommand};
+
+use maxminddb::geoip2;
+use std::net::IpAddr;
+
+use sp_core::sr25519;
+use phala_crypto::sr25519::KDF;
+use crate::secret_channel::SecretMessageChannel;
+
+const BLOCK_INTERVAL: BlockNumber = 2400;   // 8 hours
+const PROBE_DURATION: BlockNumber = 5;      // 1 min
+const IP_PROBE_URL: &str = "https://ip.kvin.wang";   // ipinfo is reported to be baned in China
+
+pub enum GeoProbeSideTaskResult {
+    Success(Geocoding),
+    Failed(String)
+}
+
+pub fn db_query_region_name<'a>(city_general_data: &'a geoip2::City) -> &'a str {
+    if let Some(city) = &city_general_data.city {
+        return *city.names.as_ref().unwrap().get(&"en").unwrap();
+    }
+
+    if let Some(subdivisions) = &city_general_data.subdivisions {
+        return *subdivisions[0].names.as_ref().unwrap().get(&"en").unwrap();
+    }
+
+    if let Some(country) = &city_general_data.country {
+        return *country.names.as_ref().unwrap().get(&"en").unwrap();
+    } else {
+        return &"NA";
+    }
+}
+
+pub fn process_block(
+    block_number: BlockNumber,
+    egress: &Sr25519MessageChannel,
+    side_task_man: &mut SideTaskManager,
+    identity_key: &sr25519::Pair,
+    geoip_city_db: String,
+) {
+    if block_number % BLOCK_INTERVAL == 1 {
```

Comment:
Purely relying on `block_number % period == magic` may results in a wavefront of the messages from all the workers to the blockchain, which make push a heavy burden at a single time point. This can be solved by giving each worker a different offset by its identity.

A simple idea could be:

1. We derive a deterministic offset for each worker
   ```
   let worker_magic = first_32_bits(hash(worker_pubkey)) % BLOCK_INTERVAL;
   ```

2. Then here we check
   ```
   if block_number % BLOCK_INTERVAL == worker_magic {
   ```